File: my_py_pkg/my_py_pkg/number_publisher.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int16


class NumberPublisher(Node):

    def __init__(self):
        super().__init__("number_publisher")
        self.get_logger().info("Hello from " + self.get_name())
        self.counter_ = 0
        self.publisher = self.create_publisher(Int16, "numbers", 10)
        self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        self.counter_ += 1
        data = Int16()
        data.data = self.counter_
        self.publisher.publish(data)


def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisher()
    
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int16.hpp"

class NumberPublisher : public rclcpp::Node
{
public:
    NumberPublisher() : Node("number_publisher")
    {
        RCLCPP_INFO(this->get_logger(), "Hello from %s", this->get_name());

        publisher = this->create_publisher<example_interfaces::msg::Int16>("numbers", 10);

        timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                         std::bind(&NumberPublisher::timerCallback, this));
    }

private:
    
    rclcpp::Publisher<example_interfaces::msg::Int16>::SharedPtr publisher;
    void timerCallback()
    {
        auto data = example_interfaces::msg::Int16();
        data.data = 3;
        publisher->publish(data);
    }

    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
